Modernize Makefile

timkenhan opened the pull request:
related: https://github.com/gentoo/gentoo/pull/35089

zippy2 committed:
gen_configkernel: Prepare for kernel-6.6

The microcode related config options were removed in [1] and
microcode loading is enabled when support for corresponding CPU
is enabled. Basically, this is equivalent to renaming options:

  CONFIG_MICROCODE_AMD -> CONFIG_CPU_SUP_AMD
  CONFIG_MICROCODE_INTEL -> CONFIG_CPU_SUP_INTEL

Reflect this change in gen_configkernel.sh.

Bug: https://bugs.gentoo.org/913659
1: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=e6bcfdd75d53390a67f67237f4eafc77d9772056
Signed-off-by: Michal Privoznik <michal.privoznik@gmail.com>
Signed-off-by: Ben Kohler <bkohler@gentoo.org>

benkohler committed:
doc: add new cmdline options to man page

Signed-off-by: Ben Kohler <bkohler@gentoo.org>

timkenhan committed:
modernize makefile

timkenhan committed:
updated software.sh

timkenhan commented:
@robbat2 can you help review?

robbat2 commented on Makefile:
```diff
@@ -5,27 +5,61 @@ MANPAGE = genkernel.8
+	gen_worker.sh \
+	path_expander.py
+
+SOFTWARE = BCACHE_TOOLS BOOST BTRFS_PROGS BUSYBOX COREUTILS CRYPTSETUP \
+	DMRAID DROPBEAR EUDEV EXPAT E2FSPROGS FUSE GPG \
+	HWIDS ISCSI JSON_C KMOD LIBAIO LIBGCRYPT LIBGPGERROR LIBXCRYPT LVM \
+	LZO MDADM MULTIPATH_TOOLS POPT STRACE THIN_PROVISIONING_TOOLS UNIONFS_FUSE \
+	USERSPACE_RCU UTIL_LINUX XFSPROGS XZ ZLIB ZSTD
```
one entry per line please, to ease future diffs

timkenhan replied:
addressed in commit [e87c328](https://github.com/gentoo/genkernel/commit/e87c328eac10865f4976cf2f96d3a5f86bbfe88c)

robbat2 replied:
now it's all on a single line? I expected
```
SOFTWARE = \
  BCACHE_TOOLS \
  BOOST \
  ...
```

timkenhan replied:
lmao so sorry i completely misinterpreted your comment

addressed in commit [0050718](https://github.com/gentoo/genkernel/pull/56/commits/0050718347eb34b7fe36fc3cf421abf8dfa775b9)

robbat2 commented on Makefile:
```diff
@@ -5,27 +5,61 @@ MANPAGE = genkernel.8
 
+BUILD_DIR = build
+
+FINAL_DEPS = genkernel.conf \
+	gen_cmdline.sh \
+	gen_initramfs.sh \
+	gen_determineargs.sh \
+	gen_arch.sh \
+	gen_bootloader.sh \
+	gen_compile.sh \
+	gen_configkernel.sh \
+	gen_funcs.sh \
+	gen_moddeps.sh \
+	gen_package.sh \
+	gen_worker.sh \
+	path_expander.py
```
sort this list if the order is not signifcant. If it is significant, document why

timkenhan replied:
addressed in commit [e87c328](https://github.com/gentoo/genkernel/commit/e87c328eac10865f4976cf2f96d3a5f86bbfe88c)

robbat2 replied:
thanks

robbat2 commented on Makefile:
```diff
@@ -56,9 +90,9 @@ distclean: clean
 	fi ;
 
-%.8: doc/%.8.txt doc/asciidoc.conf Makefile genkernel
-	a2x --conf-file=doc/asciidoc.conf --attribute="genkernelversion=$(PACKAGE_VERSION)" \
```
i don't see the genkernel version being injected anywhere now; missing from this PR?

timkenhan replied:
I removed this due to issue of not having git repo when using dist pkg.

I can add a code that generates a file with version upon `make dist`. The manpage generation code can then check for this file as well as the git to determine the version to inject. 

Any preference on the file path?

timkenhan replied:
addressed in commit [c295e6b](https://github.com/gentoo/genkernel/pull/56/commits/c295e6b7ce4efa23d5b797e24c245e360a85c03b

lemme know if you think it needs to be done differently

robbat2 commented on Makefile:
```diff
@@ -99,3 +133,57 @@ verify-shellscripts-initramfs:
+	cp -rp modules $(DESTDIR)/$(PREFIX)/share/genkernel/
+	cp -rp netboot $(DESTDIR)/$(PREFIX)/share/genkernel/
+	cp -rp patches $(DESTDIR)/$(PREFIX)/share/genkernel/
+	cp -rp worker_modules $(DESTDIR)/$(PREFIX)/share/genkernel/
```
```
DIRS = arch defaults gkbuilds modules netboot patches worker_modules

install: all
cp -ra $(DIRS) $(DESTDIR)/$(PREFIX)/share/genkernel/
```

timkenhan replied:
thanks!

addressed in commit [e87c328](https://github.com/gentoo/genkernel/commit/e87c328eac10865f4976cf2f96d3a5f86bbfe88c)

robbat2 commented on Makefile:
```diff
@@ -99,3 +133,57 @@ verify-shellscripts-initramfs:
+install: SYSCONFDIR := $(file <$(BUILD_DIR)/SYSCONFDIR)
+install: MANDIR := $(file <$(BUILD_DIR)/MANDIR)
+install: all
+	install -d $(DESTDIR)/$(SYSCONFDIR)
+	install -m 644 $(BUILD_DIR)/genkernel.conf $(DESTDIR)/$(SYSCONFDIR)/
```
why not `install -D $src $dest` ?

timkenhan replied:
addressed in commit [e87c328](https://github.com/gentoo/genkernel/commit/e87c328eac10865f4976cf2f96d3a5f86bbfe88c)

robbat2 commented on Makefile:
```diff
@@ -99,3 +133,57 @@ verify-shellscripts-initramfs:
+$(BUILD_DIR)/genkernel: $(addprefix $(BUILD_DIR)/,$(FINAL_DEPS)) $(BUILD_DIR)/software.sh
+	cp genkernel $(BUILD_DIR)/genkernel
+
+install: PREFIX := $(file <$(BUILD_DIR)/PREFIX)
+install: BINDIR := $(file <$(BUILD_DIR)/BINDIR)
+install: SYSCONFDIR := $(file <$(BUILD_DIR)/SYSCONFDIR)
+install: MANDIR := $(file <$(BUILD_DIR)/MANDIR)
```
is there a timing issue here? when are these rules evaluated? can you make sure it's always after the files are created? maybe throw error checking here to ensure they are non-empty.

timkenhan replied:
addressed in commit [ee3045f](https://github.com/gentoo/genkernel/pull/56/commits/ee3045fe2dca73560e685d535d9a14220bee7f20)

robbat2 commented on Makefile:
```diff
@@ -99,3 +133,57 @@ verify-shellscripts-initramfs:
+
+$(BUILD_DIR)/doc/genkernel.8.txt:
+	install -d $(BUILD_DIR)/doc/
+	cp doc/genkernel.8.txt $(BUILD_DIR)/doc/genkernel.8.txt
```
you seem to mix `cp` and `install` a lot; this can be probably be make more consistent.

timkenhan replied:
addressed in commit [e87c328](https://github.com/gentoo/genkernel/commit/e87c328eac10865f4976cf2f96d3a5f86bbfe88c)

robbat2 commented on Makefile:
```diff
@@ -5,27 +5,61 @@ MANPAGE = genkernel.8
 # Add off-Git/generated files here that need to be shipped with releases
 EXTRA_DIST = $(MANPAGE) ChangeLog $(KCONF)
 
-default: kconfig man
```
keep; likely change to:
`default: all`
For non-GNU make engines if they are used

timkenhan replied:
addressed in commit [e87c328](https://github.com/gentoo/genkernel/commit/e87c328eac10865f4976cf2f96d3a5f86bbfe88c)

timkenhan committed:
address some feedback from robbat2

robbat2 reviewed (commented):
close; just those 2 last issues to clean up and then we can merge

timkenhan committed:
updated the SOFTWARE list

timkenhan committed:
ensure package version is on the manpage
- added package version on manpage generation
- added versioning on dist pkg generation
- remove manpage generation from dist

timkenhan committed:
corrected variable issues in make install

timkenhan commented:
@robbat2 that should be everything! :)

robbat2 commented:
@timkenhan  it looks good now; rebase so it doesn't contain a merge and then i'll add it onto master

timkenhan commented:
@robbat2 done!

timkenhan commented:
@robbat2 also check out this PR on gentoo repo reflecting the recent changes:

https://github.com/gentoo/gentoo/pull/35089

robbat2 commented:
@timkenhan  somehow  your rebase picked up other commits that are already on the master branch?
```
gen_configkernel: Prepare for kernel-6.6 
doc: add new cmdline options to man page 
```
I'm going to exclude them for now when I merge this

robbat2 commented:
For future commits, please also ensure you have signed-off-by trailers on each commit.
